src/cio_websocket.c

```diff
@@ -50,44 +50,57 @@ enum cio_ws_frame_type {
 	CIO_WEBSOCKET_PONG_FRAME = 0x0a,
 };
 
-static void send_frame(struct cio_websocket *s, uint8_t *payload, size_t length, enum cio_ws_frame_type frame_type, cio_buffered_stream_write_handler written_cb)
+static void send_frame(struct cio_websocket *ws, struct cio_write_buffer *payload, enum cio_ws_frame_type frame_type, bool last_frame, cio_buffered_stream_write_handler written_cb)
 {
 	uint8_t first_len;
 	size_t header_index = 2;
 
-	s->send_header[0] = (uint8_t)(frame_type | WS_HEADER_FIN);
+	size_t length = 0;
+
+	struct cio_write_buffer *element = payload;
+	for (size_t i = 0; i < payload->data.q_len; i++) {
+		element = element->next;
+		length += element->data.element.length;
+	}
+
+	uint8_t first_byte = (uint8_t)frame_type;
+	if (last_frame) {
+		first_byte |= WS_HEADER_FIN;
+	}
+
+	ws->send_header[0] = first_byte;
+
 	if (length <= WS_SMALL_FRAME_SIZE) {
 		first_len = (uint8_t)length;
 	} else if (length <= WS_MID_FRAME_SIZE) {
 		uint16_t be_len = cio_htobe16((uint16_t)length);
-		memcpy(&s->send_header[2], &be_len, sizeof(be_len));
+		memcpy(&ws->send_header[2], &be_len, sizeof(be_len));
 		header_index += sizeof(be_len);
 		first_len = WS_SMALL_FRAME_SIZE + 1;
 	} else {
 		uint64_t be_len = cio_htobe64((uint64_t)length);
-		memcpy(&s->send_header[2], &be_len, sizeof(be_len));
+		memcpy(&ws->send_header[2], &be_len, sizeof(be_len));
 		header_index += sizeof(be_len);
 		first_len = WS_SMALL_FRAME_SIZE + 2;
 	}
 
-	if (!s->is_server) {
+	if (!ws->is_server) {
 		first_len |= WS_MASK_SET;
 		uint8_t mask[4];
 		cio_random_get_bytes(mask, sizeof(mask));
-		memcpy(&s->send_header[header_index], &mask, sizeof(mask));
+		memcpy(&ws->send_header[header_index], &mask, sizeof(mask));
 		header_index += sizeof(mask);
-		cio_websocket_mask(payload, length, mask );
+		// TODO: cio_websocket_mask(payload, length, mask );
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this block of commented out code. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=c%3ACommentedCode)
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=c%3AS1135)